### PR TITLE
Update native loading of root programs

### PR DESCRIPTION
**Pull request by sanfern**
This PR updates loading and unloading of the xdp root and tc root programs from L3afd. An XDP root program as well as TC filters are attached to the interface.The process of unloading xdp and tc programs is a common one, when the programs are loaded by L3AFD.

The next PR will include the loading of other ebpf programs from L3afd.

### CODE REVIEW

File: kf/kf_windows.go [hunk 1]
```diff
@@ -55,3 +57,19 @@ func (b *BPF) ProcessTerminate() error {
 func VerifyNCreateTCDirs() error {
 	return nil
 }
+
+// AttachXDP - Attach BPF Program
+func AttachXDP(bpfRootProg *ebpf.Program, index int) error {
+	return fmt.Errorf("AttachXDP - UnSupported on windows")
```
Comment (dthaler) [inline kf/kf_windows.go:63]:
```suggestion
	return fmt.Errorf("AttachXDP - Unsupported on windows")
```

File: kf/kf_windows.go [hunk 2]
```diff
@@ -55,3 +57,19 @@ func (b *BPF) ProcessTerminate() error {
...
+}
+
+// LoadTCRootProgram - not implemented in windows
+func (b *BPF) LoadTCRootProgram(ifaceName, direction string, eBPFProgram *BPF) error {
+	// not implement nothing todo
+	return fmt.Errorf("LoadTCRootProgram - TC programs UnSupported on windows")
```
Comment (dthaler) [inline kf/kf_windows.go:69]:
```suggestion
	return fmt.Errorf("LoadTCRootProgram - TC programs Unsupported on windows")
```

File: kf/kf_windows.go [hunk 3]
```diff
@@ -55,3 +57,19 @@ func (b *BPF) ProcessTerminate() error {
...
+}
+
+// UnLoadTCProgram - Remove TC filters
+func (b *BPF) UnLoadTCProgram(ifaceName, direction string) error {
+	return fmt.Errorf("UnLoadTCProgram - TC programs UnSupported on windows")
```
Comment (dthaler) [inline kf/kf_windows.go:74]:
```suggestion
	return fmt.Errorf("UnloadTCProgram - TC programs Unsupported on windows")
```
nit: s/UnLoad/Unload/ throughout

File: kf/bpf.go [hunk 4]
```diff
@@ -1098,6 +1120,90 @@ func (b *BPF) VerifyMetricsMapsVanish() error {
...
+	b.ProgMapID, ok = ebpfInfo.ID()
+	if !ok {
+		return fmt.Errorf("fetching map id failed %v", err)
+	}
+
+	return nil
+}
+
+// UnLoadProgram - Unload the program
+func (b *BPF) UnLoadProgram(ifaceName, direction string) error {
```
Comment (dthaler) [inline kf/bpf.go:1175]:
```suggestion
// UnloadProgram - Unload the program
func (b *BPF) UnloadProgram(ifaceName, direction string) error {
```

File: kf/bpf.go [hunk 5]
```diff
@@ -1098,6 +1120,90 @@ func (b *BPF) VerifyMetricsMapsVanish() error {
...
+	_, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		log.Fatal().Msgf("lookup network iface %q: %s", ifaceName, err)
```
Comment (dthaler) [inline kf/bpf.go:1178]:
```suggestion
		log.Fatal().Msgf("look up network iface %q: %s", ifaceName, err)
```
("look up" is a verb.  "lookup" is a noun.)

File: kf/kf_unix.go [hunk 6]
```diff
@@ -194,3 +202,176 @@ func VerifyNCreateTCDirs() error {
...
+
+	var parent uint32
+	if direction == models.IngressType {
+		parent = tc.HandleMinIngress
+	} else if direction == models.EgressType {
+		parent = tc.HandleMinEgress
+	}
+
+	progFD := uint32(bpfRootProg.FD())
+	bpfFlag := uint32(0x1)
```
Comment (dthaler) [inline kf/kf_unix.go:294]: what does the number 0x1 actually mean here?  Preferably make it a symbol.  Especially just in case the value might vary by platform (Linux, Windows, whatever)

File: kf/kf_unix.go [hunk 7]
```diff
@@ -194,3 +202,176 @@ func VerifyNCreateTCDirs() error {
...
+
+	return nil
+}
+
+// UnloadTCProgram - Remove TC filters
+func (b *BPF) UnloadTCProgram(ifaceName, direction string) error {
+
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		log.Fatal().Msgf("lookup network iface %q: %s", ifaceName, err)
```
Comment (dthaler) [inline kf/kf_unix.go:328]:
```suggestion
		log.Fatal().Msgf("look up network iface %q: %s", ifaceName, err)
```

File: kf/kf_unix.go [hunk 8]
```diff
@@ -194,3 +202,176 @@ func VerifyNCreateTCDirs() error {
...
+		Parent:  core.BuildHandle(tc.HandleRoot, tc.HandleMinIngress),
+	})
+
+	if err != nil {
+		log.Warn().Msgf("Could not get filters for interface \"%s\" direction %s ", ifaceName, direction)
+		return fmt.Errorf("could not get filters %v", err)
+	}
+
+	progFD := uint32(bpfRootProg.FD())
+	bpfFlag := uint32(0x1)
```
Comment (dthaler) [inline kf/kf_unix.go:353]: What does 0x1 mean here?

**Review by strcloud — APPROVED** [review]
- API Documentation for this change is necessary
 - Release note explaining  added capability, bugs and the missing features plan for next milestone.

File: kf/kf_unix.go (hunk 8 above)
Comment (sanfern) [reply kf/kf_unix.go]: Flags defined by the kernel for the BPF filter, value set is direct attach.

**Review by dthaler — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

File: kf/bpf.go [hunk 9]
```diff
@@ -243,6 +253,19 @@ func (b *BPF) Stop(ifaceName, direction string, chain bool) error {
...
 
 	if len(b.Program.CmdStop) < 1 {
+		// Program ref handle - loaded from l3afd
+		if b.CollectionRef != nil {
+			if err := b.UnloadProgram(ifaceName, direction); err != nil {
+				return fmt.Errorf("BPFProgram %s unload failed with error: %v", b.Program.Name, err)
+			}
+			log.Info().Msgf("%s => %s direction => %s - program is unloaded/detached successfully", ifaceName, b.Program.Name, direction)
+			return nil
+		} else if len(b.Program.ObjectFile) > 0 {
```
Comment (jniesz) [inline kf/bpf.go:263]: Will this ever get called?  When would we have an ObjectFile defined, but not a CollectionRef?  This would essentially mean we didn't perform a "LoadCollection" call

File: kf/bpf.go [hunk 10]
```diff
@@ -243,6 +253,19 @@ func (b *BPF) Stop(ifaceName, direction string, chain bool) error {
 	stats.SetWithVersion(0.0, stats.NFRunning, b.Program.Name, b.Program.Version, direction, ifaceName)
 
 	if len(b.Program.CmdStop) < 1 {
+		// Program ref handle - loaded from l3afd
+		if b.CollectionRef != nil {
```
Comment (jniesz) [inline kf/bpf.go:257]: I think this should be above/outside the if block.  If there is a CollectionRef defined we should unload it as I believe this means we loaded the eBPF program directly and didn't run any external binary to do this.  In this case wouldn't b.Program.CmdStop be irrelevant?

**Comment by jniesz** [comment]
PR needs a lot more detail in the description and list if this is part 1 of N PRs for R2 program loading changes.

File: kf/bpf.go [hunk 11]
```diff
@@ -1098,6 +1121,93 @@ func (b *BPF) VerifyMetricsMapsVanish() error {
...
+	}
+
+	if b.Program.ProgType == models.TCType {
+		if err := b.UnloadTCProgram(ifaceName, direction); err != nil {
+			log.Warn().Msgf("UnloadTC program failed iface %q direction %s error - %v", ifaceName, direction, err)
+		}
+	}
+
+	if b.CollectionRef != nil {
+		b.CollectionRef.Close()
```
[the remark this answers is not in the document]
Comment (sanfern) [reply kf/bpf.go:1192]: This CollectionRef.Close will unload and close all the map handles, instead doing one by one.

File: kf/bpf.go [hunk 12]
```diff
@@ -242,7 +256,17 @@ func (b *BPF) Stop(ifaceName, direction string, chain bool) error {
 	// Setting NFRunning to 0, indicates not running
 	stats.SetWithVersion(0.0, stats.NFRunning, b.Program.Name, b.Program.Version, direction, ifaceName)
 
+	// First preference to unload/stop from l3afd
+	if b.ProgMapCollection != nil {
+		if err := b.UnLoadProgram(ifaceName, direction); err != nil {
+			return fmt.Errorf("BPFProgram %s unload failed with error: %v", b.Program.Name, err)
```
Comment (jniesz) [inline kf/bpf.go:262]: Can we add interface in the error and logging?

File: kf/bpf.go [hunk 13]
```diff
@@ -1098,6 +1108,110 @@ func (b *BPF) VerifyMetricsMapsVanish() error {
...
+	return nil
+}
+
+// UnLoadProgram - Unload the program
+func (b *BPF) UnLoadProgram(ifaceName, direction string) error {
+	_, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		log.Fatal().Msgf("lookup network iface %q: %s", ifaceName, err)
+	}
+
```
Comment (jniesz) [inline kf/bpf.go:1171]: Are we planning to support stopping user space program along with the eBPF program here?

Comment (sanfern) [reply kf/bpf.go:1171]: yes, in the next PR, when we load ebpf_program and start user program combination, will add stopping user program.

File: CODEOWNERS [hunk 14]
```diff
@@ -1,3 +1,3 @@
 # Default Code Owners
 
-* @sanfern @charleskbliu0 @jniesz @dthaler
+* @sanfern @charleskbliu0 @jniesz @dthaler @ajayR006
```
Comment (jniesz) [inline CODEOWNERS:3]: This doesn't belong in this PR, please move it to a separate one.

File: kf/bpf.go (hunk 13 above)
Comment (jniesz) [reply kf/bpf.go:1167]: Can we add a TODO line then?

File: kf/kf_unix.go [hunk 15]
```diff
@@ -194,3 +202,176 @@ func VerifyNCreateTCDirs() error {
...
+
+	tcfilts, err := b.TCFilter.Get(&tc.Msg{
+		Family:  unix.AF_UNSPEC,
+		Ifindex: uint32(iface.Index),
+		Handle:  0x0,
+		Parent:  core.BuildHandle(tc.HandleRoot, tc.HandleMinIngress),
+	})
+
+	if err != nil {
+		log.Warn().Msgf("Could not get filters for interface %s direction %s ", ifaceName, direction)
```
[the remark this answers is not in the document]
Comment (dthaler) [reply kf/kf_unix.go:338]: Why was this marked as resolved?  I didn't see a response

File: kf/kf_unix.go (hunk 8 above)
Comment (dthaler) [reply kf/kf_unix.go:343]: Please add a comment or constant in the code.

File: kf/kf_unix.go (hunk 7 above)
Comment (dthaler) [reply kf/kf_unix.go:318]: Why was this marked as resolved?  I didn't see any change.

File: kf/bpf.go (hunk 13 above)
Comment (sanfern) [reply kf/bpf.go:1173]: Done

File: kf/bpf.go [hunk 16]
```diff
@@ -1098,6 +1106,123 @@ func (b *BPF) VerifyMetricsMapsVanish() error {
...
+	b.ProgMapID, ok = ebpfInfo.ID()
+	if !ok {
+		return fmt.Errorf("fetching map id failed for xdp program %s to interface %s : %v", b.Program.Name, ifaceName, err)
+	}
+
+	return nil
+}
+
+// UnloadProgram - Unload or detach the program from the interface and close all the program resources
+// TODO: Before unloading make sure user program is stopped to avoid any errors
```
Comment (dthaler) [inline kf/bpf.go:1167]: Is there a github issue filed to track this TODO?

**Review by dthaler — APPROVED** [review]

File: kf/bpf.go (hunk 16 above)
Comment (sanfern) [reply kf/bpf.go:1167]: This particular pull request is solely focused on loading root programs. However, in the next pull request, there will be a more structured flow for loading xdp and tc programs, starting with user programs and vice versa.

Comment (dthaler) [reply kf/bpf.go:1167]: Is there a github issue filed to track that work?_https://github.com/l3af-project/l3afd/blob/main/docs/CONTRIBUTING.md says "It is recommended prior to working on a PR to submit an issue in github for the change you want to make describing the change and context around it.  ..."

**Review by jniesz — APPROVED** [review]
LGTM
